poetapp/node
# feature: implement POST /files endpoint

[PR Process](https://github.com/poetapp/documentation/blob/master/process/pr-review.md#pr-process) - [PR Review Checklist](https://github.com/poetapp/documentation/blob/master/process/pr-review.md#pr-review-checklist)

### Release
 Semantic release is enabled for this repository. Make sure you follow the right commit message convention. 
 We're using semantic-release's default — [Angular Commit Message Conventions](https://github.com/angular/angular.js/blob/master/DEVELOPERS.md#-git-commit-guidelines).

## Description of Changes

closes #611 

This pr ended up being larger than needed because we don't have any kind of feature toggles in place #685.

Add the ability for users to post files to IPFS through the node.

- [x] store the resulting hash in the database
- [x] handle multiple file uploads
- [x] use multipart
- [x] use streams
- [x] returns an array of objects that contain a hash
- [x] use configuration for ipfs url


## Out of Scope / Todo after this PR
- throw error if no files passed in via multipart
- recheck to see if we can pass through the ReadStream without saving a temp file.  Tried this originally in #682 , but would like to double check to see if its possible.
- adjust functional tests to use the file endpoint
- publish messages that contain the hashes of the uploaded files
- handle larger file uploaded with other uploaded techniques than multipart ( thx @krobi64 )
  - IPFS currently only takes multipart if I remember correctly, but they were working on a better way for larger files.
  - https://gist.github.com/CMCDragonkai/6bfade6431e9ffb7fe88#a-note-about-nodejs
  - https://stackoverflow.com/questions/25810051/filereader-api-on-big-files/25813769#25813769

## Manual Test Directions
* terminal 1 ```cp docker-compose.override.yml.exampl docker-compose.override.yml```
* edit docker-compose.override.yml and uncomment poet-node's `volumes` and `command`, save.
* terminal 1 ```docker-compose up -d```
* terminal 1 ```docker-compose exec poet-node bash```
* terminal 1 ```docker-compose npm i && npm run build && npm run start```
* terminal 2 ```docker-compose exec mongo bash```
* terminal 2 ```mongo --username poet --password poetPass poet```
* terminal 3 ```echo 'my file' > text.txt```
* terminal 3 ```curl -F file=@text.txt "http://localhost:18080/files"```   
   should result in something like:  [{"hash":"QmRDfwqbJocLAEdxeVjLbUeXPKpNcj3n9GrPUvukbvfaF5"}]
* terminal 2 ```db.files.find()```   
  should result in something like: { "_id" : ObjectId("5bda22c59cb490008e6abb86"), "hash" : "QmRDfwqbJocLAEdxeVjLbUeXPKpNcj3n9GrPUvukbvfaF5" }
